Optimisation of Recrusive Fibonacci series (#41)
Developer-Students-Club-MAIT/cool_code_snippets: snippets/c++/Recursive Fibonacci in C++ with Memoization/fibonacci.md

# Recursive Fibonacci in C++ with Memoization

## Description
In an ordinary recursive function of the Fibonacci same function with the same parameter is called many times, so there are excessive recursive calls, but with memorization, we take an array of suitable size and initialize it with -1 and store the result of the unique recursive call in an array instead of -1. So when we need to call a function again, we don't call we directly take its result from the array we have already stored and if does not exist then the function is called recursively.  

## Code
```C++
#include<bits/stdc++.h>

using namespace std;
int F[10]; //Globally declared  or Static Array to store the value of called function
//It prevent from excessive recursion
int mfib(int n) {
  if (n <= 1) {
    F[n] = n;
    return n;
  } else {
    if (F[n - 2] == -1)
      F[n - 2] = mfib(n - 2);
    if (F[n - 1] == -1)
      F[n - 1] = mfib(n - 1);
    F[n] = F[n - 2] + F[n - 1];

    return F[n];
  }
}

int main() {
  int i;
  for (i = 0; i < 10; i++)
    F[i] = -1;
  printf("%d \n", mfib(7));
  return 0;
}
```

## Input

> ```7```

Reviewer: I/O blocks are not formatted correctly. The number needs to be in the second line and in the third line the closing line, similar to how we define docstrings:
```python
"""
content here
"""
```

> ```7```

Reviewer 2: ~~~suggestion
```
7
```
~~~

## Output

> ```13```

Reviewer 2: ~~~suggestion
```
13
```
~~~

## Contributed By
|Name|Github Username|College Name|
|---|---|---|
|Manikant Rai|Manikant25|SRM Institute of Science And Technology|

> 

Reviewer: remove this extra line

# Recursive Fibonacci in C++ with Memoization

## Description

> In an ordinary recursive function of the Fibonacci same function with the same parameter is called many times, so there are excessive recursive calls, but with memorization, we take an array of suitable size and initialize it with -1 and store the result of the unique recursive call in an array instead of -1. So when we need to call a function again, we don't call we directly take its result from the array we have already stored and if does not exist then the function is called recursively.  

Reviewer 2: ```suggestion
In an ordinary recursive function of the Fibonacci same function with the same parameter is called many times, so there are excessive recursive calls, but with memoization, an array of suitable size is taken and each value is initialized with -1. The result of the unique recursive call in an array is then stored replacing -1. So, when the function needs to be called again, it isn't called, instead, the result is directly taken from the array already stored and if it does not exists then the function is called recursively.
```

Reviewer 2: Also add it in the description that this code works for the first 100 numbers in the fibonacci sequence.

## Code
```C++
#include<bits/stdc++.h>

using namespace std;

> int F[10]; //Globally declared  or Static Array to store the value of called function

Reviewer 2: Change this to extend to 100 numbers in the sequence

//It prevent from excessive recursion
int mfib(int n) {
  if (n <= 1) {
    F[n] = n;
    return n;
  } else {
    if (F[n - 2] == -1)
      F[n - 2] = mfib(n - 2);
    if (F[n - 1] == -1)
      F[n - 1] = mfib(n - 1);
    F[n] = F[n - 2] + F[n - 1];

    return F[n];
  }
}

int main() {
  int i;

>   for (i = 0; i < 10; i++)

Reviewer 2: ```suggestion
  for (i = 0; i < 100; i++)
```

    F[i] = -1;

>   printf("%d \n", mfib(7));

Reviewer 2: So, I noticed that you aren't taking any input here. Can you update the program to take an input with a prompt here and also reflect that in the input block below?

  return 0;
}
```

## Input
```
7
```

## Output
```
13
```

> ## Contributed By

Reviewer 2: ~~~suggestion
```

## Contributed By
~~~

Author: Done

# Recursive Fibonacci in C++ with Memoization

> 

Reviewer: Can you describe this program in brief here? Maybe 2-3 lines?

Author: I have described the program above already

Reviewer: I can't see in this file?

Author: okay I add

Author: Sir after making changes I will have to send pull request again and close this?

> # Code

Reviewer: Make it level 2 heading (##)

```C++
#include<bits/stdc++.h>
using namespace std;
int F[10]; //Globally declared  or Static Array to store the value of called function
           //It prevent from excessive recursion
int mfib(int n)
{
 if(n<=1)
 {

>  F[n]=n;

Reviewer: The whole code needs to be formatted properly, proper indentation

 return n;
 }
 else
 {
 if(F[n-2]==-1)
 F[n-2]=mfib(n-2);
 if(F[n-1]==-1)
 F[n-1]=mfib(n-1);
 F[n]=F[n-2]+F[n-1];
 
 return F[n];
 }
}

int main()
{
 int i;
 for(i=0;i<10;i++) //Initialising array F with -1
 F[i]=-1;

 printf("%d \n",mfib(5));
 return 0;
}
```

> # Input

Reviewer: - Level 2 heading
- the input block is not ended

```7```

> # Output

Reviewer: level 2
